### ISSUE TITLE
cannot import name 'open' from 'smart_open'

### ISSUE THREAD

**Issue by littleyee** [RaRe-Technologies/smart_open#489]
I am receiving the error 
  File "C:\ProgramData\Anaconda2\lib\site-packages\gensim\utils.py", line 45, in <module>
    from smart_open import open

ImportError: cannot import name open
I am using python 2.7.16, the gensim is in 3.8.2 and smart-open is 1.10.1. Any ideas of what is going on?

**Comment by mpenkov** [comment]
What is the output of:

```
import smart_open
dir(smart_open)
```

Here is the output I see on my system (different to yours) as an example:

```
$ python
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import smart_open
>>> dir(smart_open)
['__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', 'bytebuffer', 'compression', 'concurrency', 'constants', 'doctools', 'gcs', 'hdfs', 'http', 'local_file', 'logger', 'logging', 'open', 'parse_uri', 'register_compressor', 's3', 's3_iter_bucket', 'smart_open', 'smart_open_lib', 'ssh', 'transport', 'utils', 'version', 'webhdfs']
>>> 
```

**Comment by littleyee** [comment]
The output is Python 2.7.16 (v2.7.16:413a49145e, Mar  4 2019, 01:37:19) [MSC v.1500 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import smart_open
>>> dir(smart_open)
['__all__', '__builtins__', '__doc__', '__file__', '__name__', '__package__', '__path__', '__version__', 'bytebuffer', 'doctools', 'hdfs', 'http', 'logger', 'logging', 'open', 'register_compressor', 's3', 's3_iter_bucket', 'smart_open', 'smart_open_lib', 'ssh', 'version', 'webhdfs']
>>>

**Comment by mpenkov** [comment]
Interesting. The `open` function is in the list of functions offered by the `smart_open` package. I'm not sure why it's not importing.

I originally suspected this is some kind of Python 2 issue, but it works in a Py2 environment here without any problems:

```
$ python
Python 2.7.17 (default, Apr 15 2020, 17:20:14) 
[GCC 7.5.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from smart_open import open
>>> import smart_open
>>> smart_open.__version__
'1.10.1'
>>>
```

**Comment by jcatanza** [comment]
I just installed smart_open on my  Windows 10 machine. It doesn't contain a module called "open"
(base) C:\Users\cross-entropy>python
Python 3.7.7 (default, Apr 15 2020, 05:09:04) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import smart_open
>>> dir(smart_open)
['BZ2File', 'BytesIO', 'DEFAULT_ERRORS', 'IS_PY2', 'P', 'PATHLIB_SUPPORT', 'SSLError', 'SYSTEM_ENCODING', 'Uri', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'boto', 'codecs', 'collections', 'gzip', 'hdfs', 'http', 'importlib', 'io', 'logger', 'logging', 'os', 'pathlib', 'pathlib_module', 'requests', 's3', 's3_iter_bucket', 'six', 'smart_open', 'smart_open_hdfs', 'smart_open_http', 'smart_open_lib', 'smart_open_s3', 'smart_open_webhdfs', 'sys', 'urlparse', 'urlsplit', 'warnings', 'webhdfs']
>>>

**Comment by Abhishek-Prajapat** [comment]
I am not getting the open function in the smart_open directory.
`import smart_open
dir(smart_open)
['BZ2File',
 'BytesIO',
 'DEFAULT_ERRORS',
 'IS_PY2',
 'P',
 'PATHLIB_SUPPORT',
 'SSLError',
 'SYSTEM_ENCODING',
 'Uri',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 'boto',
 'codecs',
 'collections',
 'gzip',
 'hdfs',
 'http',
 'importlib',
 'io',
 'logger',
 'logging',
 'os',
 'pathlib',
 'pathlib_module',
 'requests',
 's3',
 's3_iter_bucket',
 'six',
 'smart_open',
 'smart_open_hdfs',
 'smart_open_http',
 'smart_open_lib',
 'smart_open_s3',
 'smart_open_webhdfs',
 'sys',
 'urlparse',
 'urlsplit',
 'warnings',
 'webhdfs']`

**Comment by mpenkov** [comment]
@Abhishek-Prajapat What smart_open version are you using?

**Comment by anantguptadbl** [comment]
@Abhishek-Prajapat , I am also not able to see any open in the smart_open. I am, using the following versions
**_smart_open-1.8.0
gensim-3.8.3_**

I solved it by writing my own version of glove2word2vec which is almost copy-paste

```
glove_input_file = 'glove.6B.50d.txt'
word2vec_output_file = 'glove.6B.50d.txt.word2vec'

def gloveToword2vec(glove_input_file,word2vec_output_file):
    with open(glove_input_file, 'rb') as f:
        num_lines = sum(1 for _ in f)
    with open(glove_input_file, 'rb') as f:
        num_dims = len(f.readline().split()) - 1
    with open(word2vec_output_file, 'wb') as fout:
        fout.write("{0} {1}\n".format(num_lines, num_dims).encode('utf-8'))
        with open(glove_input_file, 'rb') as fin:
            for line in fin:
                fout.write(line)
    print("Completed")

gloveToword2vec(glove_input_file, word2vec_output_file)

```

**Comment by kaitlinchaung** [comment]
I am having a similar issue, with version smart-open-2.1.0.

This is my output of the directory: 

Python 3.7.6 (default, Jan  8 2020, 13:42:34) 

>>> import smart_open
>>> dir(smart_open)
['BZ2File', 'BytesIO', 'DEFAULT_ERRORS', 'IS_PY2', 'P', 'PATHLIB_SUPPORT', 'SSLError', 'SYSTEM_ENCODING', 'Uri', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'boto', 'codecs', 'collections', 'gzip', 'hdfs', 'http', 'importlib', 'io', 'logger', 'logging', 'os', 'pathlib', 'pathlib_module', 'requests', 's3', 's3_iter_bucket', 'six', 'smart_open', 'smart_open_hdfs', 'smart_open_http', 'smart_open_lib', 'smart_open_s3', 'smart_open_webhdfs', 'sys', 'urlparse', 'urlsplit', 'warnings', 'webhdfs']

**Comment by Auburngrads** [comment]
This issue doesn't appear to have been fixed in the latest versions `smart_open`.

Both 2.0.0 and 2.1.0 state that the `smart_open.smart_open()` is deprecated in favor of `smart_open.open()` in **smart_open/smart_open_lib.py** although this doesn't seem to be what is occurring.

Also, the list of exported functions I see from `smart_open`  does not include `open`

```
dir(smart_open)

['BZ2File', 'BytesIO', 'DEFAULT_ERRORS', 'IS_PY2', 'P', 'PATHLIB_SUPPORT', 'SSLError', 'SYSTEM_ENCODING', 'Uri', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'boto', 'codecs', 'collections', 'gzip', 'hdfs', 'http', 'importlib', 'io', 'logger', 'logging', 'os', 'pathlib', 'pathlib_module', 'requests', 's3', 's3_iter_bucket', 'six', 'smart_open', 'smart_open_hdfs', 'smart_open_http', 'smart_open_lib', 'smart_open_s3', 'smart_open_webhdfs', 'sys', 'urlparse', 'urlsplit', 'warnings', 'webhdfs']
```

My immediate issues with `gensim` appear to have been fixed by updating **gensim/utils.py** to read 

````
from smart_open import smart_open
````

**Comment by piskvorky** [comment]
@Auburngrads are you sure you're using the latest smart_open?

My guess would be you have two (or more) instances of smart_open installed, perhaps in different virtual environments. And one is old (`smart_open.smart_open`) which is what the gensim environment is picking up.

Otherwise it's really weird. How did you install smart_open? Can you uninstall it, re-install the latest version again and **post here the full installation log**?

**Comment by Auburngrads** [comment]
@piskvorky possible that I would have multiple versions installed, but seems unlikely. 

Not using any virtual environments that I'm aware of and only have `smart_open` installed to my only custom environment (not found in base).  Installed using `pip` in anaconda prompt.

Removed and reinstalled `pip install smart_open`, verified version 2.1.0 is installed and have the same issue 

Installation log file is attached.
[so_log.txt](https://github.com/RaRe-Technologies/smart_open/files/4996730/so_log.txt)

**Comment by piskvorky** [comment]
Thanks. When you open the file `c:\users\aubur\appdata\local\r-mini~1\envs\r-reticulate\lib\site-packages\smart_open\smart_open_lib.py`,  do you see the line `def open(` there?

My MD5 hash of that file in release 2.1.0 is `MD5 (smart_open_lib.py) = 5fcfed617811f6b63cd8ff44df118e83`.

**Comment by Auburngrads** [comment]
I do, yes.

I haven’t found anything in this file that I can attribute to why `open` would not be available.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by piskvorky** [comment]
What does `import smart_open; print(smart_open.__file__)` say?

**Comment by Auburngrads** [comment]
It returns C:\Users\Aubur\AppData\Local\r-miniconda\envs\r-reticulate\lib\site-packages\smart_open\__init__.py

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by piskvorky** [comment]
What's up with the `r-mini~1` vs `r-miniconda`? Do both paths exist, or is some program truncating the path like that?

**Comment by Auburngrads** [comment]
The paths are the same, although anaconda prompt is truncating the printed path.  This did highlight what my be my issue, though.

I interact with Python via Rstudio through an R package called reticulate.  This is essentially a Python REPL that uses the `r-reticulate` environment by default (The non-truncated path is returned by the REPL).  Up until this I’ve never seen a difference between the behavior of the reticulate REPL and that when I interact with Python directly.

I ran `from smart_open import open`  in Python from the Anaconda Prompt -- without error. When attempted to verify this in spyder I needed to fix an issue with `qtpy`.  After uninstalling and re-installing `qtpy` (and dependencies) everything works, both in Python and in the REPL.

I can only assume that a library required to establish the REPL became corrupted and caused the error.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10
